das/manager.go

```diff
@@ -0,0 +1,175 @@
+package das
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type samplingManager struct {
+	concurrency int
+	fetchFn     func(context.Context, uint64) error
+	storeState  func(context.Context, state)
+	state       *state
+
+	//stats       atomic.Value
+	catchUpDone   int32         // indicates if all headers are sampled
+	catchUpDoneCh chan struct{} // blocks until all headers are sampled
+
+	workersWg     sync.WaitGroup
+	jobsCh        chan uint64 // fan-out jobs to workers
+	resultCh      chan result // fan-in sampling results from worker to coordinator
+	discoveryCh   chan uint64 // receives all info about new headers discovery
+	storeCh       chan state  // communicates with backgroundStore routine
+	storeInterval time.Duration
+
+	coordinatorDone chan struct{}
+	cancel          context.CancelFunc
+}
+
+func newSamplingManager(concurrency int,
```

Comment:
It needs to be in constructor because different tests use diferent concurrency values and can run in parallel. On the other hand for application it takes value from pkg level constant that defined `daser.go`. 

BTW it is not exposed to user but could be put in config in future. Concurrency level could affect sampling speed on cost or machine resources consumption so could subject of user configuration.

Reply:
I think we want to limit user configuration for this kind of thing at the moment, but definitely in the future would consider this.